### ISSUE TITLE
[Smallfix] Flix Test

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#265]
Here is a test from `TestOption`:

```
  test("zip.01") {
    val input = "def r: Option[(Int32, Bool)] = Option/zip(None, None)"
    runAnyTest(input, Value.mkNone())
  }
```

Notice that it uses `runAnyTest` and `Value`. The problem with depending on `Value` is that we want to be able to change it in the future. The problem with `runAnyTest`:

```
  def runTest(input: String, output: Int) {
    val flix = new Flix().setOptions(options).addPath("main/src/library/Option.flix").addStr(input)
    assertResult(output)(flix.solve().get.getConstant("r"))
  }
```

is that it is a small wrapper around the `Flix` object.

Unfortunately, we seem to have come up with many of those:

- runTest
- runBoolTest
- runAnyTest
- run
- runTest
- runInterceptTest
- checkModel

... aaaaaand so on. All similar, but slightly different.

What about adding an annotation `@test` and writing:

```
@test
def testNull01: Bool = Option/null(None) == true
```

The annotation simply marks the function as a test case (which must be boolean valued). If the test returns false, then the test failed. 

We could allow you to run the tests by simply passing the `--test` flag to Flix. This would be useful for local development, but less so for the entire Flix compiler. 

Fortunately, we will be able to integrate these test cases with ScalaTest since we can programatically create tests. The only downside is that we won't be able to use assertEquals and richer assertions, but only fail on the boolean condition. Moreover, debugging a single test case will be harder.

This would not solve all our testing needs. We would still have Scala tests for test cases that should trigger error message at compile- or run-time. E.g. test cases for type errors.

The tests will be limited to the functional language. 
I don't know how we could test the logic language (?).

Other issues could be parallelism, understanding why the test case failed (returned false), issues related to re-using the same Flix instance for more than one test (?).

What else?

**Comment by mhyee** [comment]
It would be useful to have some mechanism for testing the logic language.

If I'm writing some analysis, and have some small, handwritten tests, I want to check that the analysis produces the results I expect, especially if I do some refactoring later on.

Right now, we can use ScalaTest to pull out the relation from the Flix object, and then iterate over its contents, but this is a bit cumbersome. Can we do better? Does it make sense to have something within the Flix language that checks if two relations are equal (or maybe subset of)?

**Comment by magnus-madsen** [comment]
> If I'm writing some analysis, and have some small, handwritten tests, I want to check that the analysis produces the results I expect, especially if I do some refactoring later on.

If we had negation, I guess you could write:

```
rel Actual(x: Int)
rel Expected(x: Int)

Expected(21). 
Expected(42).

false :- !Actual(x), Expected(x).
```

Is that readable to you? Unfortunately the rule does not have a name (unlike a test case), so when it fails, it will just fail with that line number.

**Comment by magnus-madsen** [comment]
Another idea is to use flixtest purely for the library (and programs written in Flix), but not necessarily for the compiler tests.

**Comment by mhyee** [comment]
> If we had negation, I guess you could write:
> 
> ```
> rel Actual(x: Int)
> rel Expected(x: Int)
> 
> Expected(21). 
> Expected(42).
> 
> false :- !Actual(x), Expected(x).
> ```
> 
> Is that readable to you? Unfortunately the rule does not have a name (unlike a test case), so when it fails, it will just fail with that line number.

Yes, this is readable, since I can just define the `Expected` relation and facts in a separate file.

If the error message prints out the actual rule that failed, then that's good enough as the "name." Of course, if we really wanted to go this route of making it part of the test framework (maybe a `@test` annotation on the rule?) we could have a counter of passed/failed tests, a summary at the end, etc.

> Another idea is to use flixtest purely for the library (and programs written in Flix), but not necessarily for the compiler tests.

This is actually what I had in mind when I read the "Flix Test" title. Having a test framework in Flix to test the Flix implementation seems a little weird.

**Comment by magnus-madsen** [comment]
Let us consider this test case (from `TestBackend`):

```

  test("Expression.Lambda.07") {
    val input =
      """def f(x: Int8, y: Int8): Int8 = (x: Int8) - y
        |def g(x: Int8): Int8 = x * 3i8
        |def h(x: Int8): Int8 = g(x - 1i8)
        |def x: Int8 = let x = 7i8; f(g(3i8), h(h(x)))
      """.stripMargin
    val t = new Tester(input)
    t.runTest(Value.mkInt32(-42), "x")
  }
```

How would we go about changing this to a Flix test?

At first glance we might try:

```
def f(x: Int8, y: Int8): Int8 = (x: Int8) - y
def g(x: Int8): Int8 = x * 3i8
def h(x: Int8): Int8 = g(x - 1i8)

@test
def ExpressionLambda07: Bool = let x = 7i8; f(g(3i8), h(h(x))) `assertEq` -42i8
```

but this has the problem of putting `f` and `g` and `h` into the current namespace. 

Probably we want to re-use those for different things. 

A potential solution would be namespaces:

```
namespace Expression.Lambda07 {
  def f(x: Int8, y: Int8): Int8 = (x: Int8) - y
  def g(x: Int8): Int8 = x * 3i8
  def h(x: Int8): Int8 = g(x - 1i8)

  @test
  def test: Bool = let x = 7i8; f(g(3i8), h(h(x))) `assertEq` -42i8
}
```

The only ugly thing is what to call the test function. Here I called it just `test`. I could also have called it `Lambda07` (at the risk of repeating myself). This will generate Scala test names: `Expression.Lambda07/test` or `Expression.Lambda07/Lambda07`.

Another option, if/when we have proper nested functions would be:

```
@test
def ExpressionLambda07: Bool = {
  def f(x: Int8, y: Int8): Int8 = (x: Int8) - y;
  def g(x: Int8): Int8 = x * 3i8;
  def h(x: Int8): Int8 = g(x - 1i8);
  let x = 7i8; 
    f(g(3i8), h(h(x))) `assertEq` -42i8
}
```

which is probably the cleanest.

**Comment by magnus-madsen** [comment]
Merged in #283
